### PR TITLE
feat: implement approx_median and approx_count_distinct where possible

### PR DESCRIPTION
**Pull request by cpcloud**
- test: ignore where argument when irrelevant
- test: thread where argument where relevant
- test: hide or remove unused arguments
- test: ensure backend approx_nunique test doesn't fail
- refactor(sqlalchemy): centralize reduction compilation
- refactor(sqlalchemy): hide unused arguments
- refactor(sqlalchemy): rewrite reduction to use method
- refactor(sqlalchemy): rewrite count distinct to use generalized reduction
- feat(sqlalchemy): implement approx_count_distinct as count distinct
- refactor(postgres): use alchemy backend approx_nunique implementation
- refactor(postgres): use base backend aggs
- refactor(postgres): hide unused arguments
- refactor(postgres): remove special random function
- test: enable approx median test for duckdb
- feat(duckdb): implement approx_count_distinct
- feat(duckdb): implement approx_median
- docs: add docstrings to `approx_nunique` and `approx_median`
- docs: better docstrings and type annotations
- test(sqlite): fix test
- feat(pyspark): implement approx_count_distinct
- test(pyspark): enable approx_median
- feat(pyspark): implement approx_median

### CODE REVIEW

File: ibis/backends/base/sql/alchemy/translator.py [hunk 1]
```diff
@@ -40,12 +43,41 @@ class AlchemyExprTranslator(ExprTranslator):
 
     context_class = AlchemyContext
 
+    _bool_aggs_need_cast_to_int32 = True
+    _boolean_cast_ops = ops.Sum, ops.Mean, ops.Min, ops.Max
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/translator.py:47]: Do we really need to configure these ops per backend? If not could we use them directly in the isinstance check?

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/translator.py:47]: Some backends don't require it, though the cast is likely okay for those too.

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/translator.py:47]: Ah, and this is handled by `_bool_aggs_need_cast_to_int32` anyway, so yeah I'll inline them.

File: ibis/backends/postgres/registry.py [hunk 2]
```diff
@@ -696,13 +667,7 @@ def _day_of_week_name(t, expr):
         ops.ExtractMillisecond: _millisecond,
         ops.DayOfWeekIndex: _day_of_week_index,
         ops.DayOfWeekName: _day_of_week_name,
-        ops.Sum: _reduction('sum'),
-        ops.Mean: _reduction('avg'),
```
Comment (kszucs) [inline ibis/backends/postgres/registry.py:700]: Are these implemented for the base class?

File: ibis/backends/base/sql/alchemy/translator.py [hunk 3]
```diff
@@ -40,12 +43,40 @@ class AlchemyExprTranslator(ExprTranslator):
 
     context_class = AlchemyContext
 
+    _bool_aggs_need_cast_to_int32 = True
+    _has_filter_syntax = False
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/translator.py:47]: Can you add `reduction` to this config option otherwise it's easy to misinterpret.

**Review by kszucs — APPROVED** [review]
Just nits, otherwise LGTM

File: ibis/backends/postgres/registry.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/postgres/registry.py:700]: Yep.

File: ibis/backends/base/sql/alchemy/translator.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/translator.py:47]: Done.

File: ibis/backends/base/sql/alchemy/translator.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/translator.py]: Done.

**Pull request merged**